RavenDB-15747 prepare deb packaging

gregolsky opened the pull request:
On how it's supposed to work: https://github.com/ravendb/ravendb/blob/f59849b7a11f0615c9aec3c8621e7c88390541fa/scripts/linux/pkg/linux-packaging.md

I also added a config key to specify where does the Setup save the server certificate after it's been done. By default it was the working dir, which won't do in this scenario.

gregolsky committed:
RavenDB-15747 prepare deb packaging

gregolsky commented on scripts/linux/pkg/deb/assets/DEBIAN/postinst:
```diff
@@ -0,0 +1,84 @@
+    systemctl restart ravendb
+fi
+
+# CopyPalFile() workaround - 4.2 only
```
@ppekrol Do you recall any other files being copied in this manner in runtime? We need to handle this ahead of time on installation when we're `root`.

ayende replied:
I think we have zstd and sodium as well

ayende commented on scripts/linux/pkg/deb/assets/DEBIAN/changelog:
```diff
@@ -0,0 +1,4 @@
+ravendb ($RAVENDB_VERSION-1) stable; urgency=medium
+
+Please visit for full set of changes:
+https://ravendb.net/docs/article-page/5.1/csharp/start/whats-new
```
This will need updating each time

```suggestion
https://ravendb.net/docs/article-page/latest/csharp/start/whats-new
```

gregolsky replied:
This one actually depends on the version we're releasing. This is a PR to 4.2 so it should be 4.2 etc.

ayende commented on scripts/linux/pkg/deb/assets/DEBIAN/control:
```diff
@@ -0,0 +1,13 @@
+Section: database
+Priority: optional
+Depends: $DEB_DEPS 
+Architecture: amd64
```
we need one for ARM as well (32 & 64, I think)

gregolsky replied:
first need to figure out using arm docker images on x64 - I think it's called `buildx`. Anyway, yes - will log that.

ayende replied:
Why is docker invovled here?

gregolsky replied:
We built package for 18.04 on 18.04 image and test its installation there, then proceed with 20.04. Actually, you're right, I mixed those 2, but I'll need to run arm on my local to test it - but that works out-of-the-box on Docker for Desktop. 

Anyway I'll add a variable for architecture.

ayende replied:
Creating the deb can be done on WSL / etc, it is just a packaging format, not related to where / how we run it.

ayende replied:
Note that we already do the cross compliation, etc.

gregolsky replied:
Yes, but on CI we're gonna run that pipeline on Ubuntu machine on Docker.

ayende commented on scripts/linux/pkg/deb/assets/DEBIAN/control:
```diff
@@ -0,0 +1,13 @@
+Vcs-Git: https://github.com/ravendb/ravendb.git
+Description: RavenDB NoSQL Database
+ A NoSQL Database that's fully transactional. 
+ Allows 1 million reads and 150000 writes per second.
```
Do we need this?

gregolsky replied:
Some package description would be nice. That's also what's shown in `apt` when you `show` the package. Would you like something else here?

ayende replied:
Yes, send that to David :-)

ayende commented on scripts/linux/pkg/deb/assets/DEBIAN/copyright:
```diff
@@ -0,0 +1,29 @@
+# the following clauses, or change it to suit. Delete these two lines
+Files: debian/*
+Copyright: 2021 Hibernating Rhinos LTD <support@ravendb.net>
+License: GPL-2+
```
Shouldn't this be?

```suggestion
License: AGPL-3+
```

gregolsky replied:
I'm not really an expert here when it comes to licensing.. This is basically about stuff in `debian` dir - the packaging, not RavenDB.

ayende replied:
I think that this is about our license, not theirs.

gregolsky replied:
Our stuff license is above. It was just how it was in template. I can remove this `debian/*` clause and wrap everything in one - https://github.com/ravendb/ravendb/pull/11802/files/d53312532f519c94e7ec3fef786663baef2acd96#diff-82262bb7894e7f610c8c2626336d25575097bd90860600953b81b40ef1a2c2d8R7

ayende commented on scripts/linux/pkg/deb/assets/DEBIAN/changelog:
```diff
@@ -0,0 +1,4 @@
+ravendb ($RAVENDB_VERSION-1) stable; urgency=medium
+
+Please visit for full set of changes:
+https://ravendb.net/docs/article-page/5.1/csharp/start/whats-new
```
Or use:

```suggestion
https://ravendb.net/docs/article-page/$RAVENDB_VERSION/csharp/start/whats-new
```

gregolsky replied:
yeah but we'd need to cut that to first 2 numbers. Will do

ayende commented on scripts/linux/pkg/deb/assets/DEBIAN/postinst:
```diff
@@ -0,0 +1,84 @@
+chmod 550 /usr/local/ravendb/server/{Raven.Server,rvn}
+
+if command -v setcap &> /dev/null; then
+    setcap CAP_NET_BIND_SERVICE=+eip /usr/local/ravendb/server/Raven.Server
```
Need to also setup the debug permissions, no?

gregolsky replied:
yes

ayende commented on scripts/linux/pkg/deb/assets/DEBIAN/postrm:
```diff
@@ -0,0 +1,55 @@
+
+    purge)
+    
+    rm -v /etc/ravendb/settings.json
```
Remove the certificate as well?

gregolsky replied:
yes

gregolsky replied:
Actually since it's in `/var/lib/ravendb/security` it's removed a line below.

ayende commented on scripts/linux/pkg/deb/assets/DEBIAN/prerm:
```diff
@@ -0,0 +1,51 @@
+fi
+
+# 4.2 only
+rm -vf /usr/local/ravendb/server/librvnpal.so
```
zstd & sodium as well
But aren't we removing the directory entirely?

gregolsky replied:
`apt` lists the files in the package and removes them one by one. Then if the dir is empty removes that dir. If something's copied manually it will prevent directory removal - so we need to remove manually the stuff we manually added.

ayende commented on scripts/linux/pkg/deb/assets/build-deb.sh:
```diff
@@ -0,0 +1,76 @@
+with codecs.open(defaultSettingsPath, 'r', 'utf-8-sig') as f:
+    settings = json.load(f);
+    
+del settings['DataDir']
```
I think we should default to `/data/ravendb` and create that in the install phase

gregolsky replied:
We do that, create /var/lib/ravendb/data

ayende commented on scripts/linux/pkg/deb/assets/wrapper.sh:
```diff
@@ -0,0 +1,27 @@
+#!/bin/bash -x
+
+# This is a wrapper allowing to run RavenDB by regular users *wherever* (directory-wise) they like
```
You are making this exposed as `ravendb` command, which is likely not ideal.
maybe `ravendb-local` ?

gregolsky replied:
OK, or do we scrap it altogether? Please see my other comment on this one.

ayende replied:
yes

ayende commented on scripts/linux/pkg/linux-packaging.md:
```diff
@@ -0,0 +1,88 @@
+- set file permissions properly
+
+- setcap on the Raven.Server
+
```
`setcap`  to allow debugging

ayende commented on scripts/linux/pkg/linux-packaging.md:
```diff
@@ -0,0 +1,88 @@
+
+- settings - `/etc/ravendb/settings.json`
+
+- data - `/var/lib/ravendb/data`
```
Default path in `/data/ravendb` ? Should we go for the layout consistency or assume that this is the main purpose of the instance and make the data location obvious?

ayende replied:
Thinking this over, MySQL & Postgres all store their data in similar paths, so `/var/lib` it is.

gregolsky replied:
Yes

ayende commented on scripts/linux/pkg/linux-packaging.md:
```diff
@@ -0,0 +1,88 @@
+
+- data - `/var/lib/ravendb/data`
+
+- security (e.g. certificate)  - `/var/lib/ravendb/security`
```
Who can read into this? Also, do we write the secret key into this location as well or to `$HOME/.secrets` ?

gregolsky replied:
Only `root` and `ravendb` group users. Also what determines where the `.secrets` is kept? Also it it does keep it 'at home' then it's `ravendb` user's home.

ayende replied:
It's either on `$HOME/.ravendb/secret.key` or you can specify `Security.MasterKey.Path` - I think we want the later option

gregolsky replied:
So I presume we want that like the rest of security - `/var/lib/ravendb/security/secret.key`.

ayende replied:
Yes

gregolsky replied:
You can specify that if it's there. If a new one needs to be generated it'll die with "no such file" https://github.com/ravendb/ravendb/blob/42bd9f403e176081040712826fb8e7b217b1afa7/src/Raven.Server/ServerWide/SecretProtection.cs#L726

But.. if it has length 0 it'll generate it. I'll just `touch` it in the script then.

ayende commented on scripts/linux/pkg/linux-packaging.md:
```diff
@@ -0,0 +1,88 @@
+
+- rvn link - `/usr/bin/rvn -> /usr/local/ravendb/rvn`
+
+- ravendb - `/usr/bin/ravendb -> /usr/local/ravendb/Raven.Server` 
```
The ability to run this independently as a stand alone server isn't really attractive, what scenario are we trying to enable here?

ayende commented on scripts/linux/pkg/linux-packaging.md:
```diff
@@ -0,0 +1,88 @@
+
+## Running as a regular user
+
+The idea is for the users to be able to run multiple instances of RavenDB on a multi-user system, keeping the data and the settings in their home directory.
+When run as a regular user we start on a random port without settings, producing RavenData dir (and all the other dirs) in the working dir.
```
I can't really think of a good scenario for this. If user have installed us, they are likely going to be wanting to use the dameon.

gregolsky replied:
Well, I thought about things like a university setting up a multi-user server for all students and then they have the ability to run RavenDB for their purposes. Scrap it?

ayende replied:
Yes, in almost all such cases, docker containers would be better

ayende commented on scripts/linux/pkg/linux-packaging.md:
```diff
@@ -0,0 +1,88 @@
+### Installation
+
+When installed and run the first time we start in `SetupMode` `Initial`.
+
```
Need to consider this scenario. Assume that I'm using SSH to connect to the server.
How do I do the setup?

We can either make it work via command line browser (doubtful) or need to emit the right command to show SSH tunneling and connect to remotely to the local port.

gregolsky replied:
CLI browser is out of question. They are not that powerful to run Studio or even the setup part - too complex. As far as tunneling is concerned it's probably their thing to do it - we can put something in the docs and in the package readme, but I would not invest there. 

Main question here is how do we set it up when having just the terminal. Ideas for the future:

- a CLI setup wizard script (drawing some menus and all), sending Setup Wizard payloads via curl, doing LE and puff - it's there. It's feasible for sure. 

- simpler and cheaper, no wizard, but elaborate `ravendb-setup` for which you'd pass few to multiple switches (I mean options elaboration level certbot) and they would init the settings file
```bash
$ sudo ravendb-setup \
-o raven-setup-package.zip \
--license ./license.json \
--accept-eula \
--agree-le-tos \
--node-ip 1.2.3.4 \
--node-ip 2.2.3.4 \
--node-ip 1.6.6.7 \
--domain ravendb.com
# probably few more here
```
or
```
ravendb-setup --localhost --unsecured --yes-i-accept-its-unsecured
```
People are getting used to this kind of commands nowadays - mentioned `certbot` or `docker` have this kind of interface and it works for them.

Without a tool like this we can:
- start in setup as we do by default - document accessing it the first time if set up on a "headless" machine
- start unsecured on localhost, which is not ideal

Thoughts?

ayende replied:
the `ravendb-setup` command might be the best option, yes. But also needs to be interactive, I think

That said, I think that setting up the system to allow SSH tunneling is easiest to do and give us all the benefits.
We just need to provide the command and ensure that the user is aware of that.

By default, we use dynamic port, which isn't ideal, though.

gregolsky replied:
We don't have a good place to show such command when installing the package - server is not running yet and we don't know the port. Even if we did, we don't know how does the user connect to this server in the first place (the key/password, tunneling the ssh itself through some bastion) since we would need to mimic that. I don't think we're able to do that.

solutions:
- `ravendb-setup` mentioned above
- include `open-ravendb-setup-to-the-world.sh` in the package
- add explicit info in the docs on how to setup on headless/ssh-only machines - we already have some piece here sort of https://ravendb.net/docs/article-page/5.1/Csharp/start/installation/setup-examples/aws-linux-vm
- not use dynamic port for setup? (still can't show exact command but can print that we're gonna wait for you on `localhost:53780`

ayende replied:
Yes, we will pretty much have to run a fixed port. We can use a high one, which is unlikely to be busy.

gregolsky replied:
Added a message like this during the install:
![image](https://user-images.githubusercontent.com/1541805/112005197-d3977700-8b22-11eb-8936-4db57d5e8a18.png)

ayende replied:
Please add the `ssh` command to do the port forwarding as well here.

gregolsky replied:
Since we don't know the external hostname the best I think we can do is something like this:
![image](https://user-images.githubusercontent.com/1541805/112822173-e14f7e00-9087-11eb-9fea-1f78ccbd277a.png)

Does it work for you?

ayende replied:
Can't we use: `dig +short myip.opendns.com @resolver1.opendns.com ` ?

gregolsky replied:
OK, sure. Might cover most cases. Here's what I have now:
```
$ sudo dpkg -i ravendb_4.2.112-0_amd64.deb
Selecting previously unselected package ravendb.
(Reading database ... 70102 files and directories currently installed.)
Preparing to unpack ravendb_4.2.112-0_amd64.deb ...
Unpacking ravendb (4.2.112-0) ...
Setting up ravendb (4.2.112-0) ...
Adjust createdump binary permissions...
Adjust Raven.Debug binary permissions...
###
#
#  Please navigate to http://127.0.0.1:53700 in your web browser to complete setting up RavenDB.
#  If you set up the server through SSH, you can tunnel RavenDB setup port and proceed with the setup on your local.
#
#  For public address:   ssh -N -L localhost:8080:localhost:53700 ubuntu@54.198.78.116
#
#  For internal address: ssh -N -L localhost:8080:localhost:53700 ubuntu@ip-10-0-2-76
#
###
Processing triggers for libc-bin (2.27-3ubuntu1) ...
```

Sufficient? Still don't know if somebody logged in with password or key so we might be wrong with regards to auth method.

ayende replied:
Looks really nice, yes.

ayende replied:
It is enough that we provide this, the actual auth is on them.

We can't handle everything.

For example, we might not be able to handle it if they force to bind to localhost and use another port, etc. 
Not on us in this case.

gregolsky committed:
RavenDB-15747 allow to specify where would the setup wizard save the certificate file

gregolsky committed:
RavenDB-15747 add information on tunneling the setup

gregolsky committed:
RavenDB-15747 adjust postinst ssh tunneling script message

gregolsky committed:
RavenDB-15747 applied adjustments after Pawel's review
- add arm/v7 build profile script
- remove all references to 5.0
- get dotnet version from runtime.txt
- do not depend on dotnet-runtime-deps, instead extract dependencies
  from it and depend on that
- make copyright year a variable
- scripts cleanup

gregolsky commented:
Applied Paweł's suggestions. I think it's ready to be merged - further work will come in 5.1 - there's `amd64` profile to add and probably some script tweaks to be made.

ayende commented on scripts/linux/pkg/deb/assets/build.sh:
```diff
@@ -0,0 +1,81 @@
+        export RAVEN_SO_ARCH_SUFFIX="linux.x64"
+        ;;
+
+    raspberry-pi)
```
what about arm 64?

gregolsky replied:
will come with 5.1 - unsupported in 4.2

ayende commented on scripts/linux/pkg/deb/assets/ravendb/debian/ravendb.service:
```diff
@@ -0,0 +1,26 @@
+Environment="RAVEN_Indexing_NugetPackagesPath=/var/lib/ravendb/nuget"
+Environment="RAVEN_Logs_Path=/var/log/ravendb/logs"
+Environment="RAVEN_Security_AuditLog_FolderPath=/var/log/ravendb/audit"
+Environment="RAVEN_Security_MasterKey_Path=/etc/ravendb/security/master.key"
+Environment="RAVEN_Setup_Certificate_Path=/etc/ravendb/security/server.pfx"
```
Who owns this directoyr?

gregolsky replied:
root:ravendb

gregolsky replied:
```
/etc/ravendb:                                                
total 8                                                      
drwxrwx--- 2 root ravendb 4096 Apr 21 13:00 security         
-rw-rw---- 1 root ravendb  239 Apr 21 13:00 settings.json    
                                                             
/etc/ravendb/security:                                       
total 0                                                      
-rw-rw---- 1 root ravendb 0 Apr 21 13:00 master.key          
```
